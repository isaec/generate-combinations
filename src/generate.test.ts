import { expect, it, describe } from "vitest";
import {
  generate,
  generateTemplate,
  one,
  optional,
  some,
  Value,
} from "./generate";

generate<{
  opt?: number;
  required: number;
  alsoOptional?: number[] | string;
  numberArray?: Array<1 | 2 | 3> | "test";
}>({
  opt: optional(1),
  required: one([5, 6]),
  alsoOptional: optional([1]),
  numberArray: some([1, 2, 3]),
});

const genTest = <T extends Record<string, Value>>(
  obj: generateTemplate<T>,
  log?: boolean
) => [obj, generate<T>(obj, log)];

it.each([
  genTest({
    a: some([1, 2]),
    c: 6,
  }),
  genTest<{
    a: number[];
    l?: number;
    b: number;
    c: number;
  }>({
    a: some([1, 2]),
    l: optional(5),
    b: 10,
    c: 6,
  }),
  genTest<{
    one: number;
    optional?: 5;
  }>({
    one: one([1, 2]),
    optional: optional(5),
  }),
])(`matches snapshots for %s`, (_template, generateObject) => {
  expect(generateObject).toMatchSnapshot();
});

// 1048576 combinations
// console.log(
//   generate<{}>({
//     a: some([1, 2, 3, 4]),
//     b: some([1, 2, 3, 4]),
//     c: some([1, 2, 3, 4]),
//     d: some([1, 2, 3, 4]),
//     e: some([1, 2, 3, 4]),
//   }).length
// );
