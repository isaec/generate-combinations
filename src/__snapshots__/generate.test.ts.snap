// Vitest Snapshot v1

exports[`matches snapshots for { a: [Combination], c: 6 } 1`] = `
[
  {
    "a": [],
    "c": 6,
  },
  {
    "a": [
      1,
    ],
    "c": 6,
  },
  {
    "a": [
      2,
    ],
    "c": 6,
  },
  {
    "a": [
      1,
      2,
    ],
    "c": 6,
  },
]
`;

exports[`matches snapshots for { a: [Combination], l: [Combination], b: 10, c: 6 } 1`] = `
[
  {
    "a": [],
    "b": 10,
    "c": 6,
    "l": 5,
  },
  {
    "a": [
      1,
    ],
    "b": 10,
    "c": 6,
    "l": 5,
  },
  {
    "a": [
      2,
    ],
    "b": 10,
    "c": 6,
    "l": 5,
  },
  {
    "a": [
      1,
      2,
    ],
    "b": 10,
    "c": 6,
    "l": 5,
  },
  {
    "a": [],
    "b": 10,
    "c": 6,
  },
  {
    "a": [
      1,
    ],
    "b": 10,
    "c": 6,
  },
  {
    "a": [
      2,
    ],
    "b": 10,
    "c": 6,
  },
  {
    "a": [
      1,
      2,
    ],
    "b": 10,
    "c": 6,
  },
]
`;

exports[`matches snapshots for { one: [Combination], optional: [Combination] } 1`] = `
[
  {
    "one": 1,
    "optional": 5,
  },
  {
    "one": 2,
    "optional": 5,
  },
  {
    "one": 1,
  },
  {
    "one": 2,
  },
]
`;
